### DISCUSSION TITLE
Third-party plugin and RevitLookup.Snoop

### DISCUSSION THREAD

**Discussion by ricaun** [jeremytammik/RevitLookup#182] [Ideas]
Hello,

I was experimenting to make it possible for my plugin to open RevitLookup with my custom object, so I can snoop and debug or something like that.

I add the following code in the main RevitLookup code so my application easily could use reflection and snoop my custom class.

```C#
using Microsoft.Extensions.DependencyInjection;
using RevitLookup.Core.Objects;
using RevitLookup.Services.Contracts;

namespace RevitLookup;

public static class Snoop
{
    public static void Navigate(object value)
    {
        NavigateObject(new SnoopableObject(null, value));
    }
    public static void Navigate(Type type)
    {
        NavigateObject(new SnoopableObject(type));
    }
    public static void Navigate(Autodesk.Revit.DB.Document document, object value)
    {
        NavigateObject(new SnoopableObject(document, value));
    }
    public static void NavigateObject(SnoopableObject value)
    {
        RevitLookup.Application.Raise(() =>
        {
            var window = Host.GetService<IWindow>();
            window.ShowAttached();
            window.Scope.GetService<ISnoopService>()!.Snoop(value);
        });
    }
}
```

And I'm using the code below in my plugin.

```C#
public static class RevitLookup
{
    private static Type GetRevitLookupSnoopType()
    {
        var revitLookup = AppDomain.CurrentDomain.GetAssemblies()
            .FirstOrDefault(e => e.GetName().Name.Contains("RevitLookup"));
        var type = revitLookup.GetType("RevitLookup.Snoop");
        return type;
    }

    public static void Snoop(object value)
    {
        var type = GetRevitLookupSnoopType();
        if (type is not null)
        {
            var navigate = type.GetMethod("Navigate", new[] { typeof(object) });
            navigate.Invoke(null, new[] { value });
        }
    }
}
```

Here is a video: https://youtu.be/fApLw-kaOOM

🙂

**Comment by jeremytammik** [comment]
Thank you! That looks great! Can it be more fully automated? Otherwise, publishing this will require precise documentation to make it completely fool-proof... I can add it to the blog as you describe it... is this using the current release 2024.0.8? I am writing a blog post to announce the release, and I could easily add this note as well.

**Reply by ricaun** [reply]
The class `RevitLookup.Snoop` does no exist in the main version, in my computer I running a custom version the that class.

**Comment by jeremytammik** [comment]
aha. would you like to (a) wait for @Nice3point to say what he thinks and (b) add a pull request to integrate your custom version?

**Comment by Nice3point** [comment]
I'll think about an API option based on @ricaun and @CADBIMDeveloper  wishes https://github.com/jeremytammik/RevitLookup/issues/171 

We need to make an API that will cover different uses, I'll get to it this weekend or next week

**Comment by Nice3point** [comment]
@ricaun We can add an API to the NUGET package to avoid using reflection

**Reply by Nice3point** [reply]
@jeremytammik  Jeremy, can Autodesk give you a Token to publish a package to Nuget on their behalf, or should we use our own account?

**Comment by jeremytammik** [comment]
No, I do not have the possibility to publish anything officially as Autodesk. Furthermore, I have been asking for at least five years to publish RevitAPI.dll and RevitAPIUI.dll on NuGet. Everybody I discuss this with says, yes, sure, good idea. However, nothing happens. Hundreds of other developers are forced to publish their own unofficial copies. Occasionally, someone at Autodesk complains about that, and question whether it is legal. Snafu, totally. Sorry, but that seems to be the way it is.

**Reply by Nice3point** [reply]
Yeah, it's really weird that autodesk doesn't support Nuget at all. I have a fully automated Nuget packaging and posting system. if autodesk needs help automating their processes I can help

On the legality side they have a rather strange approach, the whole if it wasn't for third party packages, autodesk would have lost a lot of developers  🙈

**Comment by ricaun** [comment]
> @ricaun We can add an API to the NUGET package to avoid using reflection

Yes, the package probably should be better and other features could be added as well, but is more work and another project to maintain...

In this case, just opening the Snoop Navigate gonna work fine with reflection, I kinda prefer to have a utility class without any strong reference to some dll.

I never used a plugin to talk to another as a service like that, I know Revit Api there is a way to create ExternalService but never tested and don't know if fits. I could add the RevitLookup.dll as a reference and right now gonna work. 😆

**Reply by Nice3point** [reply]
I've never worked with ExternalService, I'll have to try it out)

**Reply by ricaun** [reply]
No, looks like the ExternalService is not for that case.

If you wanna enter this rabbit hole: https://thebuildingcoder.typepad.com/files/sd10752_revit_external_services_arnost_lobel_handout.pdf

**Comment by jeremytammik** [comment]
Yes, I think external services are for Revit BIM specific functionality.
